packages/workflow/src/utils.ts

```diff
@@ -10,23 +16,26 @@ export const deepCopy = <T>(source: T, hash = new WeakMap(), path = ''): T => {
 	if (hash.has(source)) {
 		return hash.get(source);
 	}
-	// Date
-	if (source instanceof Date) {
-		return new Date(source.getTime()) as T;
-	}
 	// Array
 	if (Array.isArray(source)) {
 		clone = [];
 		const len = source.length;
-		for (i = 0; i < len; i++) {
-			clone[i] = deepCopy(source[i], hash, path + `[${i as string}]`);
+		for (let i = 0; i < len; i++) {
+			clone[i] = deepCopy(source[i], hash, path + `[${i}]`);
 		}
 		return clone;
 	}
+	// Date and other Serializable objects
+	const toJSON = (source as Serializable).toJSON;
+	if (typeof toJSON === 'function') {
+		// eslint-disable-next-line @typescript-eslint/no-unsafe-call
+		return toJSON.call(source) as T;
+	}
+
```

Comment:
I still don't get it, here we are copying an object and not serializing
`deepCopy` should just copy and if you want JSON string then after copying just call `JSON.stringify`
or just leave out copying and call  `JSON.stringify` immediately

I feel here we want to mix the 2 in 1

Reply:
the issue here is that we are trying to build a semantically correct object cloning, which unfortunately breaks stuff as what we had before this wasn't being always used as a true object cloning mechanism. for example, we started cloning dates in `deepCopy`, but all of the existing code expects dates to be be ISO strings in the cloned objects.

Reply:
I see, in this case wouldn't be just enough modification to return dates as string?

Instead of this:
```ts
if (source instanceof Date) {
  return new Date(source.getTime()) as T;
}
```

Just this:
```ts
if (source instanceof Date) {
  return source.toJSON() as T;
}
```

Also with adding a TODO comment to change it back once the rest of the code is updated

Reply:
unfortunately, to be backward compatible, we need to retur `obj.toJSON()` for any object that has that method, not just `Date`.

I can add a note and a TODO.

Reply:
Then actually just checking if that method exists on `source` and call it, this could be just another condition under the part where the primitives returned

Reply:
I'd propose the only changes which can be seen in the following screenshot (compared to master), what do you think?
![image](https://user-images.githubusercontent.com/5410822/199500337-936297aa-4052-49dc-a4c7-dd7663d58012.png)

Reply:
not 100% what you mean. but, I've move the `toJSON` block to be right after the primitives block. If that's now what you meant, can you please share a snippet?

Reply:
Yes I meant exactly that, and I just fixed the types afterwards but if you have other solution I'm fine with that
The point is calling `.toJSON` after the primitives check